tables/spaces.uti: Fixed typos and two broken replacement rules

hammera opened the pull request:
Hi Chris and Bert,

My colleague, Norbert Markus detected and fixed few typo errors and two broken space replacement rules into tables/spaces.uti file, based the Liblouis documentation (the replace opcode part). The \x2060 and \xfeff characters not have hashmark character before comment part, and Liblouis put the first word of the rule comment the Braille output.

I show a simple example the reproduction the tables/hu-hu-g1.ctb table (spaces.uti included the chardefs.cti).
I used the tools/lou_translate tables/unicode.dis,tables/hu-hu-g1.ctb command, and typed the al\x2060ma word in Linux terminal.

Look the wrong example output, and what output show the lou_trace command:
```tools/lou_trace  unicode.dis,hu-hu-g1.ctb
alm⁠a
⠁⠇⠍⠺⠕⠗⠙⠁
1.	lowercase	a	1
2.	lowercase	l	123
3.	lowercase	m	134
4.	replace		12357-12347-257-37
5.	lowercase	a	1
```

You see, before m character the unicode Braille output Liblouis put the word part the text, this is not right. Because if I understanding right the replace operation documentation, if the second part is empty (and perhaps have a tab character and follow up a comment part), Liblouis need simple cut off the \x2060 character the Braille output).

Look the right example:
```
tools/lou_trace  unicode.dis,hu-hu-g1.ctb
al⁠ma
⠁⠇⠍⠁
1.	lowercase	a	1
2.	lowercase	l	123
3.	replace		
4.	lowercase	m	134
5.	lowercase	a	1
```

Look the \xfeff character wrong unicode Braille output:
```
lou_trace unicode.dis,hu-hu-g1.ctb
al﻿ma
⠁⠇⠣⠑⠗⠕⠍⠁
1.	lowercase	a	1
2.	lowercase	l	123
3.	replace		2457-137-257-12347
4.	lowercase	m	134
5.	lowercase	a	1
`

Right result after Norbert fix (Norbert fix skyp right the zero word the unicode Braille output when I trying test the \xfeff unicode character replacement rule):
```tools/lou_trace  unicode.dis,hu-hu-g1.ctb
al﻿ma
⠁⠇⠍⠁
1.	lowercase	a	1
2.	lowercase	l	123
3.	replace		
4.	lowercase	m	134
5.	lowercase	a	1
```

Not only hungarian language affected this issue related, lot of tables use the spaces.uti file. If this is possible, please add this issue the 3.25 milestone, and merge this pull request the Liblouis master branch.

I verifyed make distwin command, the operation ran succesfully into the Docker container. I verifyed all language tests the tests directori with make -j8 check command, all tests ran succesfully, except xfail marked tests. So, I think not will be have Braille table regression after merge operation.

News file bug fix entry suggestion if merged Norbert change:
>tables/spaces.uti: Fixed typos and two broken replacement rules
>
>Thanks Norbert Markus the broken replacement rules detection 
>and proper fixes based the Liblouis documentation.

Thanks the cooperation,

Attila

hammera commented:
Bert, Chris, how you seeing:
Have chance Norbert spaces.uti related small bug fix to fitting the 3.25 release cicle?
I need doing a testcase file for this bug fix related? If yes, I fast hammering a little testcase file this request related.
If fit the release time period and you not have other larger reviewing related works with other languages table improvements related, please review this bug fix and add this small bug fix PR the 3.25 release cicle.
Attila

egli commented:
Excellent catch! Thank you @hammera . Maybe a small test case would be nice. Maybe add it to an existing test file. Isn't there `tests/braille-specs/spaces.yaml`? Would that be the right place?

hammera commented:
@egli, the tests/braille-specs/spaces.yaml is an excellent place, because uses more tables in the test file when ran the spaces related testcases. Thanks the nice suggestion.
What tables would like you the tests? I not known for example africaans, irish, dutch, german languages, but for example en-us-comp8.ctb, en-us-g1.ctb, en-us-g2.ctb, en_US.tbl, hu-hu-comp8.ctb, hu-hu-g1.ctb, hu-hu-g2.ctb tables related testcases is good?
Hungarian tables related testcases not have the spaces.yaml, oldest time added the hu-hu-comp8_harness.yaml and hu-hu-g1_harness.yaml.

I will be showing a little diff output for english testcases related before committing, to prewent unneed reviewing related work. The hungarian tables related testcases extending is easyer.

I will Hammering a little diff patch to you see perhaps how looks the new test cases

bertfrees commented:
Like Christian said, a test would be nice. The purpose of spaces.yaml is just to record the current behavior, to notify us when something changes. So it is not needed to actually know the languages. If a test is more substantiated it should be moved to another YAML file.

One way to test this change is to add `- ['\x2060', '']` and `- ['\xfeff', '']` to every existing `testmode: forward` section in spaces.yaml. Another way that is both more thorough and requires less work is to apply one test to all the tables listed in the extra/generate-display-names/display-names file. Like this:

```yml
table: afr-za-g1.ctb
table: afr-za-g2.ctb
table: ar.tbl
table: ar-ar-g2.ctb
table: ar-ar-comp8.utb
...
flags: {testmode: forward}
tests:
  - ['\x2060', '']
  - ['\xfeff', '']
```

The tables that don't pass the test (e.g. because they don't include spaces.uti) could then be removed from the list.

hammera commented:
Bert, understanding your sintax suggestion. But in extra/generate-display-names/display-names file default I see a 201 element list, with contains a * character, a table filename path and a textual label what the purpose of the table.
Now, I not yet see this file any test case part, my understanding previous I thing this file is only a table describe file.

How can I running this test case file to verify the right result after I doing the modification?
Enough the make check command in tests directory, or need running the make command in extra/generate-display-names directory to I see right test result with all tables related?

Sorry if I asking a trivial thing, never I previous need doed fix this file yet, but it's never too late to learn a practical and new modern way of bulk board testing. :-):-)

If need running the make command in extras/generate-display-names directory, what go language version need installing? Enough the golang-go 1.13 version from Ubuntu apt repository (I using Ubuntu 20.04 my development box), or need the latest 1.19.5 snap version?
The .github/workflows/metadata.yml simple installs golang package before running the generate display names workflow part into metadata.yml file.

If we finalized the diff output before I committing the modification, we do squash commit, or separate test case related commit? :-):-)
usual you favorite this squash commit technique (me too), but if you would like a separate new commit entry this testcase extending related, please tell me. 

The advantage of this is that you can more easily undo the modification related to the test case extension if you want to move the bulk testing to a new file. At the moment, I cannot yet estimate how many lines will be added to the display-name file in relation to mass testing.

Attila

bertfrees commented:
My suggestion was just to create a test and apply it to the list of tables contained in "display-names" files, just because this file happens to list all the top-level tables. It's like an index. I can do the change by the way. I just explained the idea here, not expecting that you should do it.

hammera commented:
Ok, thanks.
Doing the modification.

Attila

hammera commented:
Ok, partially done the modification, If I see right a small test basis, forward direction is OK.

The backward or bothDirections mode not working this two characters related. I will show a test result later.

Spaces.uti containing following two rules the \x2060 and \xfeff characters related:
replace \x2060		#WORD JOINER
replace \xfeff		#ZERO WIDTH NO-BREAK SPACE (also Unicode BOM)

If I understanding this two rules right, the \x2060 and \xfeff characters it should simply be ignored by Liblouis when back-translating, since the second part of the replace operator code does not contain any characters.

Ok, I get following test results with I don't no yet how can possible resolve, except we not using only the forward mode the bulk testing part.
```
Cannot parse input string.
lou_checkyaml:tests/braille-specs/spaces_bulk.yaml:14: Failure
Table: afr-za-g1.ctb
Display table: checkyaml_inline_table_at_line_1

Cannot parse input string.
lou_checkyaml:tests/braille-specs/spaces_bulk.yaml:14: Failure
Table: ar.tbl
Display table: checkyaml_inline_table_at_line_1

Cannot parse input string.
lou_checkyaml:tests/braille-specs/spaces_bulk.yaml:14: Failure
Table: ar-ar-g2.ctb
Display table: checkyaml_inline_table_at_line_1

Cannot parse input string.
lou_checkyaml:tests/braille-specs/spaces_bulk.yaml:15: Failure
Table: afr-za-g1.ctb
Display table: checkyaml_inline_table_at_line_1

Cannot parse input string.
lou_checkyaml:tests/braille-specs/spaces_bulk.yaml:15: Failure
Table: ar.tbl
Display table: checkyaml_inline_table_at_line_1

Cannot parse input string.
lou_checkyaml:tests/braille-specs/spaces_bulk.yaml:15: Failure
Table: ar-ar-g2.ctb
Display table: checkyaml_inline_table_at_line_1
```
I absolute sure doing a sintax error, but don't no what. :-):-)
The 14. line containing following element the test case file:
```
  - ['', '\x2060']```
```
We put noback prefix before the two replace rules into spaces.uti file? I readed the automated test cases documentation, but not founded a solution this sintax error related. Possible I begin little tireding. :-):-)

Look the example test case file, temporary I finalizing the forward direction until we not founding a solution the backward or bothDirections mode, because I am absolute sure forward direction will be ok:
```
display: |
  include unicode-without-blank.dis
  display 9 9
  display a a
table: afr-za-g1.ctb
table: ar.tbl
table: ar-ar-g2.ctb
flags: {testmode: forward}
tests:
  - ['\x2060', '']
  - ['\xfeff', '']
flags: {testmode: backward}
tests:
  - ['', '\x2060']
  - ['', '\xfeff']
```

Attila

hammera commented:
Fantastic, all checks passed, not break anything when I doed the new testcases.
I added two forward direction bulk testcase files.
My local machine when I looked the make check result, all testcase files passed before I doed the push operation.
Don't surprise, I need cut two file part the testcase files, because when bulk testing happening (more table lines defined with one test section), lou_checkyaml enable only 150 table file/test section.
I surprised this, because I don't known this information.
So, I added all tables the testcase files if using the spaces.uti file.

Bert, don't surprise, I doed now single commit to you easyer review new change.
If all changes good, and need, we squashing when the final merge happening.

Please review the new change, hopefully you not need change lot of things.

Attila

bertfrees commented on tests/braille-specs/spaces_bulk_test_part1.yaml:
```diff
@@ -0,0 +1,176 @@
+  - ['\x2060', '']
+  - ['\xfeff', '']
+
+# Test two special word situation in hu-hu-g1.ctb table.
+# The first example not need separate the n and y letter the braille
+# output if a word containing the \xfeff simbol.
+# The second example if the word containing the \xfeff special
+# character, need separate the n and y letter the brailleoutput.
+# This method use a hungarian language Braille free converter if the
+# user need define an exception with a word dictionary.
+table: hu-hu-g1.ctb
+flags: {testmode: forward}
+tests:
+# This is the normal example when not need separate the n and y letter
+  - ["Anya", '⠨⠁⠫⠁']
+# This is the second example when general need separate the n and y letter
+  - ["An﻿ya", '⠨⠁⠝⠽⠁']
```
Can we move this to one of the Hungarian test files?

bertfrees commented on tests/braille-specs/spaces_bulk_test_part1.yaml:
```diff
@@ -0,0 +1,176 @@
+# Bulk testing international tables two spaces.uti related
+# modifications (\x2060 and \xfeff character replacement)
+# Because the Yaml system enable only 150 table/yaml file, need
+# separate this tests with two parts.
```
I suggest we increase `MAXTABLES` to something higher than 154.

Also, I would add this whole test to spaces.yaml.

hammera commented:
Bert, you wrote:
„Can we move this to one of the Hungarian test files?”
Yes, of course, thanks the suggestion.
I removed this two entry temporary the spaces test.
But first in hungarian math greek letters related pull request you need reviewing the hungarian test files rename related commit to I put back the two special hungarian tests a hungarian yaml file after spaces.uti related modification and hungarian test file rename related modification is merged. :-):-)

Perhaps I put this two special testcases future the tests/braille-specs/hu-hu-g1_special_consonants.yaml test file after this pull request are approve and merge, because this test file not affected the rename related in hungarian greek letters related pull request.

I  Hammered the tests/braille-specs/spaces.yaml.
I doed my local branch the test file merge related task, and increased in tools/lou_checkyaml.c the MAXTABLES integer value from 150 to 400.

400 tables/yaml file is enough your openion?
If yes, I commits the change and you final approve this pull request if not need doing any change.

I now going lunch and drink a coffee, because in friday I little tire.

Attila

hammera commented:
Bert, please review the tools/lou_checkyaml.c related commit and tests/braille-specs/spaces.yaml related commit before I or you squash the changes if not have another remaining cosmetic change.
All checks passed right.

Attila

hammera commented:
Hi Boys,

If have lot of work the 3.25 release related, and the extended spaces.yaml related test case approve is not fit the remaining time (the last two commits are not reviewed yet), possible merge before release time the spaces.uti related modifications only (partial merge), and future merge the extended spaces.yaml test case and lou_checkyaml.c related commit?
I see already have 12 or 13 opened another pull requests (possible another bigger pull requests have more work), because Since some few pull request a check not ran succesfull and needs fixup (the vietnamese table improvement, the spanish greek letters related pull requests), possible this changes related will be more works.
With future only need reviewing the lou_checkyaml.c related modification, and review the extended tests/spaces.yaml. The approved spaces.uti related modification is ready, not will be change (frozen).

I don't want to seem worried, but tomorrow I have an online meeting where we have to report to my boss whether our proposed changes will be included in the next release (change related to Hungarian Greek letters, related to the spaces.uti subtable error correction, hungarian hyphenation table related fix, etc). :-):-)
So, at the meeting, I should be able to answer whether the finalization of these repairs will fit in the approx. In 10 days as you have other merge requests with you perhaps to need deal with.

I need to work a bit more on a possible other merge request (the extended_braille_input branch related, I added new forward and backward translations and new test cases), I will probably be able to submit this on Saturday. This modification will be containing few critical important update, needs land the 3.25 release. The hu-hu-g1_special_consonant.yaml related future modification will be not need reviewing, I added only new changed exceptions to the testcase file.
Of course, the future modification rans succesfull the big hungarian corpus test.

The modification will not cause a regression error, it only affects the Hungarian tables, not other tables.

I can then help with other merge requests afterwards if needed? Other merge requests related if the changed source have another repository, I have possibility to create diff output to online checks running succesfull (the spanish and wietnamese tables related). Perhaps this will also help the authors to successfully move forward and have their changes made into the 3.25 release.

Attila

egli commented:
Hey @hammera we love your PRs and we will do our best to get them into the release

hammera commented:
Ok Chris, thanks your answer. I know, you and Bert have lot of work before marc 6, so I waits patient. :-):-)

Many thanks the cooperation, and the lot of patience If I have perhaps unneed questions. :-):-)

Attila

bertfrees reviewed (approved):
Great stuff. Thanks for taking care of it @hammera.

hammera commented:
Hy Chris,

Thank you the pull request merge, and the cooperation.

Attila
